### ISSUE TITLE
MP Reactive Messaging Kafka Connector: Authentication - SASL.OAUTHBEARER

### ISSUE THREAD

**Issue by hutchig** [OpenLiberty/open-liberty#7759]
Not in scope. 

https://docs.confluent.io/current/kafka/authentication_sasl/authentication_sasl_oauth.html#kafka-oauth-auth
"and is only suitable for use in non-production Kafka installations"

**Comment by hutchig** [comment]
Gordon Hutchison   [May 29th at 9:52 AM]
My squad is building a feature “MicroProfile Reactive Messaging” https://bigblue.aha.io/features/API-27 Part of this feature is to build a CDI annotation that provides for a sort of message driven bean (@Incoming/@Outgoing).
As part of the first release we will provide a means to read/write Kafka messages so part of our job is to map *from Liberty idiomatic security config to what the Kafka client (or IBM Event Streams) expects*.  We would like to get the approach/design we use to be reviewed by some of the WAS dev security SME’s - whom might be good candidates for this?

15 replies

YK Chang   [8 days ago]
It will play nicely with MP JWT and EE Security too, right?  I know @Chunlong Liang is our contact for the former while @areddy is for the latter.

Alasdair Nottingham   [8 days ago]
I’d suggest starting with @jmulvey since he is the security architect

Gordon Hutchison   [8 days ago]
Thanks @alasdair and @yeekangc I will go to Jim and cc Chunlong and Ajay initially. In terms of ‘playing nicely’ I would define that to be just ‘what a user would expect’ and what would make them happy, I meant by ‘Liberty idiomatic’ all security standards/specs that feed into how Liberty behaves. Some things that Kafka CAN support they no longer recommend using for ‘real’ data and we will have a  ordered priority list of Kafka security protocols that aims to match what IBM Event Streams (Cloud and ICP) uses. (edited)

John Alcorn   [8 days ago]
one of the issues I hit, with JWT and an MDB, is that there's no JWT to propagate to a downstream microservice.  I had to drop down to IBM proprietary (com.ibm.ws.*) SPIs to create the JWT, which I'm not happy with.  Seems like a similar thing would happen with a Kafka "MDB".  usually you either drive an authentication from a UI, or propagate a JWT from the caller - but there's no "caller" (that knows about a JWT) in the MDB case

John Alcorn   [8 days ago]
I briefly considered stuffing the JWT in a JMS header, but that's no good if the message doesn't get processed immediately (JWTs expire)

Alasdair Nottingham   [8 days ago]
are you trying to propagate identity with the message or just wanting to be able to use JWT to make outbound calls from message driven work?

John Alcorn   [8 days ago]
the latter is what I ended up doing

Chunlong Liang   [8 days ago]
https://docs.confluent.io/current/kafka/authentication_sasl/authentication_sasl_oauth.html

Chunlong Liang   [8 days ago]
On client side, I would expect that we implement org.apache.kafka.common.security.auth.AuthenticateCallbackHandler, either self-issue a JWT, or fetch a JWT (or opaque access_token) from a 3rd party OAuth(or OIDC server), or propagate existing JWT . (edited)

Chunlong Liang   [8 days ago]
On server side, we implement org.apache.kafka.common.security.auth.AuthenticateCallbackHandler to securely validate JWT (local introspection) or opaque access_token (remote introspection).

Chunlong Liang   [8 days ago]
As for token expiration due to waiting on client side, wonder if callback handler should refresh access_token if token is about to expire

Chunlong Liang   [8 days ago]
For Client side implementation of org.apache.kafka.common.security.auth.AuthenticateCallbackHandler, maybe it is common to fetch a JWT/access_token from a dedicated trusted OIDC server  in production environment . For example, in ICP cluster, you may want to fetch token from ICP's platform authentication service (an OIDC implementation). The self-issue JWT may not be common in production environment

Chunlong Liang   [8 days ago]
For example, in ICP cluster, the kafta client can either use access_token already available to ICP console (accessible via cookie),  or registers itself as oauth client to ICP's security service and request its own access_token. (edited)

Chunlong Liang   [8 days ago]
If client is not running under any user identity, oauth client credential grant type maybe the option if using 3rd party OAuth server.

Chunlong Liang   [8 days ago]
if client is running under user credential (user name + password), oauth resource owner password credential grant could be used.
